stm32/libcanard/bxcan/src/bxcan.c

```diff
@@ -0,0 +1,753 @@
+    // let:
+    //   BS = 1 + BS1 + BS2                                             -- Number of time quanta per bit
+    //   PRESCALER_BS = PRESCALER * BS
+    // ==>
+    //   PRESCALER_BS = PCLK / BITRATE
+    const uint32_t prescaler_bs = peripheral_clock_rate / target_bitrate;
+
+    // Searching for such prescaler value so that the number of quanta per bit is highest.
+    uint8_t bs1_bs2_sum = (uint8_t)(max_quanta_per_bit - 1U);
+
+    while ((prescaler_bs % (1U + bs1_bs2_sum)) != 0U)
+    {
+        if (bs1_bs2_sum <= 2U)
+        {
+            return false;  // No solution
+        }
+        bs1_bs2_sum--;
+    }
+
+    const uint32_t prescaler = prescaler_bs / (1U + bs1_bs2_sum);
+    if ((prescaler < 1U) || (prescaler > 1024U))
+    {
+        return false;  // No solution
+    }
+
+    // Now we have a constraint: (BS1 + BS2) == bs1_bs2_sum.
+    // We need to find such values so that the sample point is as close as possible to the optimal value,
+    // which is 87.5%, which is 7/8.
+    //
+    //   Solve[(1 + bs1)/(1 + bs1 + bs2) == 7/8, bs2]  (* Where 7/8 is 0.875, the recommended sample point location *)
+    //   {{bs2 -> (1 + bs1)/7}}
+    //
+    // Hence:
+    //   bs2 = (1 + bs1) / 7
+    //   bs1 = (7 * bs1_bs2_sum - 1) / 8
+    //
+    // Sample point location can be computed as follows:
+    //   Sample point location = (1 + bs1) / (1 + bs1 + bs2)
+    //
+    // Since the optimal solution is so close to the maximum, we prepare two solutions, and then pick the best one:
+    //   - With rounding to nearest
+    //   - With rounding to zero
+    uint8_t bs1 = (uint8_t)(((7U * bs1_bs2_sum - 1U) + 4U) / 8U);  // Trying rounding to nearest first
+    uint8_t bs2 = (uint8_t)(bs1_bs2_sum - bs1);
+    BXCAN_ASSERT(bs1_bs2_sum > bs1);
+
+    {
+        const uint16_t sample_point_permill = (uint16_t)(1000U * (1U + bs1) / (1U + bs1 + bs2));
+
+        if (sample_point_permill > MaxSamplePointLocationPermill)  // Strictly more!
+        {
+            bs1 = (uint8_t)((7U * bs1_bs2_sum - 1U) / 8U);  // Nope, too far; now rounding to zero
+            bs2 = (uint8_t)(bs1_bs2_sum - bs1);
+        }
+    }
+
+    const bool valid = (bs1 >= 1U) && (bs1 <= MaxBS1) && (bs2 >= 1U) && (bs2 <= MaxBS2);
+
+    // Final validation
+    // Helpful Python:
+    // def sample_point_from_btr(x):
+    //     assert 0b0011110010000000111111000000000 & x == 0
+    //     ts2,ts1,brp = (x>>20)&7, (x>>16)&15, x&511
+    //     return (1+ts1+1)/(1+ts1+1+ts2+1)
+    if ((target_bitrate != (peripheral_clock_rate / (prescaler * (1U + bs1 + bs2)))) || !valid)
+    {
+        // This actually means that the algorithm has a logic error, hence assert(0).
+        BXCAN_ASSERT(0);  // NOLINT
+        return false;
+    }
+
+    out_timings->bit_rate_prescaler    = (uint16_t) prescaler;
+    out_timings->max_resync_jump_width = 1U;  // One is recommended by UAVCAN, CANOpen, and DeviceNet
```

Comment:
Fixing my own code. Nothing to see here, move along please.

```suggestion
    out_timings->max_resync_jump_width = 1U;  // One is recommended by DS-015, CANOpen, and DeviceNet
```

Reply:
Nothing seen ;-)